### PR TITLE
Implement GenericCamera1 and GenericCamera2.

### PR DESCRIPTION
**Pull request by Ernovation**
All protocol implementations implement the interface StreamClient from the common package.

### CODE REVIEW

**Comment by pedroSG94** [comment]
Hello,

Thank you for your PR.

I can see multiple methods that will crash with NPE when you call it before startStream. For example setAuthorization method.
To solve this you need to create all RTMP, RTSP and SRT clients on constructors.

Also, you are creating duplicate code in GenericCamera1 and GenericCamera.
I think that create a GenericStreamClient that extends from StreamBaseClient like [RtmpStreamClient](https://github.com/pedroSG94/RootEncoder/blob/master/library/src/main/java/com/pedro/library/util/streamclient/RtmpStreamClient.kt) but that require RtmpClient, RtspClient and SrtClient is the best solution for this case.
For example, setAuthorization method could be implemented like this:
```kotlin
  override fun setAuthorization(user: String?, password: String?) {
    rtmpClient.setAuthorization(user, password)
    rtspClient.setAuthorization(user, password)
    srtClient.setAuthorization(user, password)
  }
```
And then only call connect of the required client depend of endpoint as you did.
What do you think about it?
Also, we will need to add this feature to DisplayBase, FromFileBase and OnlyAudioBase

**Comment by Ernovation** [comment]
Hello,

I agree that the GenericCamera classes should have an instance of each of the protocol client classes, and call the methods on these instances, as you propose, and call connect() on the required client.
As I am implementing this, I am not sure what to do about the classes that implement StreamBaseClient. It appears that these classes only forward the calls to the appropriate protocol specific class. But now all these protocol specific classes implement the new StreamClient interface. Would it be okay to remove the StreamBaseClient and its children, and return the class that implements StreamClient directly from the method getStreamClient() in e.g. Camera1Base? Then the Camera1Base can have a protected member of interface StreamClient, and the base can implement getStreamClient() directly. In the child class of Camera1Base this member can be set to an instance for the appropriate protocol. I think the Camera1Base will even be able to implement even more methods that are now abstract, because it has a member of StreamClient itself. In fact, perhaps all methods can be implemented directly in the CameraBase classes, and the classes like RtspCamera1 only have constructors. And this makes me wonder if we could just turn the CameraBase classes into a template class, e.g. Camera1<T>, where T can be RtspClient, RtmpClient, SrtClient or GenericClient. I think that would be really pretty. What do you think?
The only thing that should be taken care of, is the StreamClientListener, with the method onRequestKeyframe(). Is that still required? Could we solve that some other way?

I agree that this also must be done for DisplayBase, FromFileBase and OnlyAudioBase, but I first wanted to check with you about the plan, before implementing it all, and I only use the Camera2Base. :-)

I have done some of the implementation of what I propose above, but will wait for your opinion before continuing, so that I can continue in the right direction.

**Comment by pedroSG94** [comment]
Hello,

> Would it be okay to remove the StreamBaseClient and its children, and return the class that implements StreamClient directly from the method getStreamClient() in e.g. Camera1Base?

StreamBaseClient have the function to avoid users use methods that should be managed by the API like setVideoInfo or connect because this methods require a specific way to work. This is to avoid errors using the API. For example call connect without start encoders. Basically is a way to avoid duplicate code and only allow access to desired methods.
Also, it is more confuse if you want extend Camera2base without know about StreamClientBase implementation (in my opinion).

The current way force you to implement the abstract method in a non null way so you can make sure that this will work if anyone extend it for others protocols (the new protocols must have all commons methods or at least throw an exception to notify if it is not supported). With your suggestion it is not mandatory so you can let StreamClientBase without implement it since you haven't a method to force the user implement it and maybe crash with NPE if you don't initialize a StreamClientBase class.
Also, the current way keep Camera2Base totally protocol agnostic.

> I think the Camera1Base will even be able to implement even more methods that are now abstract, because it has a member of StreamClient itself. In fact, perhaps all methods can be implemented directly in the CameraBase classes, and the classes like RtspCamera1 only have constructors. And this makes me wonder if we could just turn the CameraBase classes into a template class, e.g. Camera1, where T can be RtspClient, RtmpClient, SrtClient or GenericClient. I think that would be really pretty. What do you think?

Create a template could be interested  but your suggestion force the user that currently use this library to change so much code. 
I really don't like this. I think we must keep compatibility with olders versions with minimal code changes.
Also, I don't want create an interface that must be implemented in all protocols because maybe that protocol don't support fews methods. For example:
If I add support for RTMFP (RTMP under UDP) then I need add setProtocol to RtmpClient like RstpClient but SRT can't implement this method because SRT only use UDP and TCP will never be supported. In this case force add setProtocol to SrtClient only to throw an exception feel really weird.

> The only thing that should be taken care of, is the StreamClientListener, with the method onRequestKeyframe(). Is that still required? Could we solve that some other way?

Call onRequestKeyFrame is totally necessary because all protocols need start with a keyframe to work. Maybe we can find a better way to do it but we can't skip this and set VideoEncoder directly don't feel good.

> I agree that this also must be done for DisplayBase, FromFileBase and OnlyAudioBase, but I first wanted to check with you about the plan, before implementing it all, and I only use the Camera2Base. :-)

Don't worry about it. I can do it. For now, you can implement only camera1 and/or camera2 and I will do it for the others cases. I said only to keep in mind that the idea is write a minimal code without duplicate code because we will add this to all that cases so the code should be write in a way to avoid it

In summary:
- I suggested a new StreamClientBase for Generic because this seem an easy way to avoid duplicate code.
- I don't want implement protocols code in Camera2Base to keep it protocol agnostics
- In my opinion we don't need add any code to protocols to support GenericCamera2 so we should avoid it if possible and only modify library module for this case.
- This implementation must keep compatibility with older library versions so template is discarded for now (even if it is a good idea currently, this lose sense after implement GenericCamera2 because you don't need to worry about the protocol)

**Comment by Ernovation** [comment]
Hi Pedro,

I have created a new StreamClientBase and made as few changes to the other code as possible. :-)

It would have been easier if all the protocol clients would implement the interface that is called from the Camera{1,2}Base classes, because then it would be easier to reuse code. Now there is still code in the <Protocol>Camera{1,2} classes that is copied. I would prefer to get some Client class from the Camera{1,2} class to do all the protocol specific things, and not have protocol specific Camera{1,2} classes at all. But that is then for future versions. For now I have removed the interface in the common module, which I had all the protocol classes implement.

What do you think of the new version that I pushed to my repo?

**Comment by pedroSG94** [comment]
I have no so much time today to check but this feels good. I will check it deep tomorrow when I have more time.
For now, I let you a comment with a question

**Comment by pedroSG94** [comment]
Hello,

I will merge this PR with my branch feature/refactor-commons, add a little refactor and merge with master.

Thank you for your PR
